### DISCUSSION TITLE
what is best practice to not repeat yourself when using components?

### DISCUSSION THREAD

**Discussion by boynet** [alpinejs/alpine#4072] [1. Help]
so I am in the process of rewrite some old website using bootstrap 4

and I am looking at alpine + tailwind + [alpine components](https://alpinejs.dev/components)

and it make me wonder, I need to put all this code for every element?
like if I need 10 dropdown elements, I will need to repeat the logic for each dropdown?
there is no some clever trick to write the logic once and make it work with class names or something?

**Comment by ekwoka** [comment]
Most of the time you'd use your templating to handle that.

You can use `x-bind` to bundle directives together into a set that you can apply, or use custom directives.

**Reply by boynet** [reply]
thanks, I can wrap it in a directive server side, but still the user will get the code for the dropdown logic x 10 times I am doing it right?

**Reply by ekwoka** [reply]
Why would it?

I'm not sure I understand.

**Reply by boynet** [reply]
I have drop down component like here https://alpinejs.dev/component/dropdown
its 30 lines of code

now in my server side I have dropdown directive and I can use it without repeating the code:
<x-dropdown>..a..</x-dropdown>
<x-dropdown>..b..</x-dropdown>
<x-dropdown>..c..</x-dropdown>

buy my server will actually duplicate the code for each dropdown, so if the code is 30 line x 10 its 300 lines of html just for simple dropdowns

I am missing something?

like in other framework I can have 1 place for the dropdown code:
class dropdown{}

and my server only deliver simple html with a "dropdown" class
but here the js logic is inside the html so it will get duplicated for each dropdown

**Reply by ekwoka** [reply]
Well, sure, the html your server sends is the html it sends.

If you have multiple of something on the page, it will have multiple of them on the page.

You can make it less by abstracting the directives into bindings.
